Core/include/Acts/Propagator/StepperConcept.hpp

```diff
@@ -9,215 +9,124 @@
-        };
-// clang-format on
-
-}  // namespace Concepts::Stepper
-
-template <typename stepper, typename state = typename stepper::State>
-constexpr bool StepperConcept =
-    Acts::Concepts ::Stepper::SingleStepperConcept<stepper, state>::value ||
-    Acts::Concepts ::Stepper::MultiStepperConcept<stepper, state>::value;
-template <typename stepper>
-constexpr bool StepperStateConcept =
-    Acts::Concepts ::Stepper::StepperStateConcept<stepper> ||
-    Acts::Concepts ::Stepper::MultiStepperStateConcept<stepper>;
-}  // namespace Acts
+    };
+  };
+
+  requires requires(const Stepper& s, const State& t) {
+    { s.position(t) } -> std::same_as<Vector3>;
+    { s.direction(t) } -> std::same_as<Vector3>;
+    { s.qOverP(t) } -> std::same_as<double>;
+    { s.absoluteMomentum(t) } -> std::same_as<double>;
+    { s.momentum(t) } -> std::same_as<Vector3>;
+    { s.charge(t) } -> std::same_as<double>;
+    { s.time(t) } -> std::same_as<double>;
+    { s.outputStepSize(t) } -> std::same_as<std::string>;
+
+    requires requires(const ConstrainedStep::Type st) {
+      { s.getStepSize(t, st) } -> std::same_as<double>;
+    };
+  };
+};
+
+/// @brief Concept that is satisfied by single-steppers.
+template <typename Stepper, typename State = typename Stepper::State>
+concept SingleStepper = CommonStepper<Stepper, State> &&
+    requires(const Stepper& s, State& t) {
+  requires requires(const FreeVector& fv, const BoundVector& bv,
+                    const BoundSquareMatrix& bm, const Surface& sf) {
+    { s.update(t, fv, bv, bm, sf) } -> std::same_as<void>;
+  };
+
+  requires requires(const Vector3& v1, const Vector3& v2, double d1,
+                    double d2) {
+    { s.update(t, v1, v2, d1, d2) } -> std::same_as<void>;
+    { s.getField(t, v1) } -> std::same_as<Result<Vector3>>;
+  };
+};
+
+/// @brief Concept that is satisfied by multi-steppers.
+template <typename Stepper, typename State = typename Stepper::State>
+concept MultiStepper = CommonStepper<Stepper, State> && requires {
+  // TODO for now we do not check if the ComponentProxy does fulfill a concept
+  typename Stepper::ComponentProxy;
+
+  // TODO for now we do not check if the ConstComponentProxy does fulfill a
+  // concept
+  typename Stepper::ConstComponentProxy;
+
+  requires requires(const Stepper& s, State& t) {
+    { s.numberComponents(t) } -> std::same_as<std::size_t>;
+    { s.clearComponents(t) } -> std::same_as<void>;
+    { s.removeMissedComponents(t) } -> std::same_as<void>;
+  };
+};
+
+/// @brief Concept that is satisfied by steppers.
+template <typename _Stepper, typename State = typename _Stepper::State>
+concept Stepper =
+    SingleStepper<_Stepper, State> || MultiStepper<_Stepper, State>;
+
+/// @brief Concept that is satisfied by stepper states.
+template <typename State>
+concept StepperState = requires(const State& t) {
+  { t.covTransport } -> decayed_same_as<const bool&>;
+  { t.pathAccumulated } -> decayed_same_as<const double&>;
+};
```

Comment:
The previous version was using `Concepts` as a kind of detail namespace to keep the main namespace clean. Maybe we can keep that here, and call these `StepperConcept` and `StepperStateConcept` as before?

Reply:
In the current iteration of this PR, the concepts keep living in `Acts::Concepts`. Do you mean keeping the "helper" concepts (e.g. `SingleStepper` and such) in a different namespace?

Reply:
Hmm @paulgessinger, for something like the `same_as_any_of` concept, would you prefer `Acts::same_as_any_of`, `Acts::Concepts::same_as_any_of`, `Acts::SameAsAnyOfConcept`, `Acts::Concepts::SameAsAnyOfConcept`, or something else?

Reply:
Updated as discussed.